drivers/serial/uart_nrfx_uarte.c

```diff
@@ -251,6 +323,476 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+						    NRF_TIMER_TASK_COUNT));
+
+		if (ret != NRFX_SUCCESS) {
+			return -EIO;
+		}
+#else
+		nrf_uarte_publish_set(uarte,
+				      NRF_UARTE_EVENT_RXDRDY,
+				      data->async->rx_cnt.ppi);
+		nrf_timer_subscribe_set(data->async->timer.p_reg,
+					NRF_TIMER_TASK_COUNT,
+					data->async->rx_cnt.ppi);
+
+#endif
+		ret = gppi_channel_enable(data->async->rx_cnt.ppi);
+		if (ret != NRFX_SUCCESS) {
+			return -EIO;
+		}
+	} else {
+		nrf_uarte_int_enable(uarte, NRF_UARTE_INT_RXDRDY_MASK);
+	}
+
+	nrf_uarte_int_enable(uarte,
+			     NRF_UARTE_INT_ENDRX_MASK |
+			     NRF_UARTE_INT_RXSTARTED_MASK |
+			     NRF_UARTE_INT_ERROR_MASK |
+			     NRF_UARTE_INT_ENDTX_MASK |
+			     NRF_UARTE_INT_TXSTOPPED_MASK |
+			     NRF_UARTE_INT_RXTO_MASK);
+	nrf_uarte_enable(uarte);
+	k_timer_init(&data->async->rx_timeout_timer, rx_timeout, NULL);
+	k_timer_user_data_set(&data->async->rx_timeout_timer, dev);
+	k_timer_init(&data->async->tx_timeout_timer, tx_timeout, NULL);
+	k_timer_user_data_set(&data->async->tx_timeout_timer, dev);
+
+	return 0;
+}
+
+static int uarte_nrfx_tx(struct device *dev, const u8_t *buf, size_t len,
+			 u32_t timeout)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	if (!nrfx_is_in_ram(buf)) {
+		return -ENOTSUP;
+	}
+
+	if (data->async->tx_buf || data->async->tx_size) {
+		return -EBUSY;
+	}
+	data->async->tx_buf = buf;
+	data->async->tx_size = len;
+	nrf_uarte_tx_buffer_set(uarte, buf, len);
+	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTTX);
+	if (data->uart_config.flow_ctrl == UART_CFG_FLOW_CTRL_RTS_CTS) {
+		k_timer_start(&data->async->tx_timeout_timer, timeout, 0);
+	}
+	return 0;
+}
+
+static int uarte_nrfx_tx_abort(struct device *dev)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	if (data->async->tx_buf == NULL) {
+		return -EFAULT;
+	}
+	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STOPTX);
+
+	if (data->uart_config.flow_ctrl == UART_CFG_FLOW_CTRL_RTS_CTS) {
+		k_timer_stop(&data->async->tx_timeout_timer);
+	}
+
+	return 0;
+}
+
+static int uarte_nrfx_rx_enable(struct device *dev, u8_t *buf, size_t len,
+				u32_t timeout)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	if (hw_rx_counting_enabled(data)) {
+		nrfx_timer_clear(&data->async->timer);
+	} else {
+		data->async->rx_cnt.cnt = 0;
+	}
+	data->async->rx_total_byte_cnt = 0;
+	data->async->rx_total_user_byte_cnt = 0;
+
+	data->async->rx_timeout = timeout;
+	data->async->rx_timeout_slab = MAX(timeout / RX_TIMEOUT_DIV,
+					   MS_PER_TICK);
+	data->async->rx_size = len;
+	data->async->rx_buf = buf;
+	data->async->rx_offset = 0;
+	nrf_uarte_rx_buffer_set(uarte, buf, len);
+
+	nrf_uarte_event_clear(uarte, NRF_UARTE_EVENT_ENDRX);
+	nrf_uarte_event_clear(uarte, NRF_UARTE_EVENT_RXSTARTED);
+
+	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTRX);
+	data->async->rx_enabled = true;
```

Comment:
Move this before triggering the `STARTRX` task. Otherwise, you may end up servicing `endrx_isr` without this field being set.